### DISCUSSION TITLE
`remote: Bypassed rule violations for refs/heads`

### DISCUSSION THREAD

**Discussion by zaikunzhang** [community/community#51047] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

Since two days ago, I have been receiving the following message when pushing to my repo at https://github.com/libprima/prima (note the `remote: xyz` parts). 

```
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 16 threads
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 478 bytes | 478.00 KiB/s, done.
Total 4 (delta 2), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (2/2), completed with 2 local objects.
remote: Bypassed rule violations for refs/heads/main:
remote: 
remote: - Commits must have valid signatures.
remote: 
remote: - Waiting on code owner review from USER_NAME.
remote: 
To github.com:libprima/prima.git
   8d9987c3..85a1f0d0  main -> main
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
```

Remark:

-  I am the only owner, administrator, and contributor of the repo. It may be true that I am bypassing some branch protection rules set by myself, but the same warning has never been emitted before --- I think it was suppressed by default. 

- This is probably related to https://github.com/orgs/community/discussions/50878 .
Note that I have done the following according to https://github.blog/2023-03-23-we-updated-our-rsa-ssh-host-key/ , but this does not solve the problem. 
```
$ ssh-keygen -R github.com
$ curl -L https://api.github.com/meta | jq -r '.ssh_keys | .[]' | sed -e 's/^/github.com /' >> ~/.ssh/known_hosts
```

Is someone experiencing this same? How to get rid of the `remote: - Commits must have valid signatures.` and `remote: - Waiting on code owner review from USER_NAME.` ? 

Thanks.

**Comment by airtower-luna** [comment]
That message doesn't have anything to do with SSH keys, it's warning you that your push ignored a branch protection rule. I assume that you have admin rights to the repository and the rule is configured to allow admins to ignore it.

**Reply by zaikunzhang** [reply]
Thank you @airtower-luna . I see it in the same way as you. 

However, it has never happended before the SSH key incedent of GitHub. In addition, [someone else](https://github.com/orgs/community/discussions/50878#discussioncomment-5428612) experiences something similar due to the change of the SSH key. That is why I thought they were related. 

> I assume that you have admin rights to the repository and the rule is configured to allow admins to ignore it.

Right. **I am the only owner, administrator, and contributor of the repo.** The warning is not wrong, but it seems to have been suppressed by default before. The question is, how and why is it turned on now (starting two days ago)? 

Thanks. 

Remark: I found three similar cases on the Internet.

- GitHub Community Discourse: https://github.com/orgs/community/discussions/50878#discussioncomment-5428612
- StackExchange: https://stackoverflow.com/questions/75839192/github-at-least-1-approving-review-is-required-by-reviewers-with-write-access 
- GitHub code: https://github.com/RikaShiro/FYP22070/blob/a15eae1e9a0cdbb1b2d7c14d640549ada6c8f17a/FYP/posts/20230320-20230326.txt#L15

All of them started to occur two days ago. I speculate that they are caused by some recent changes on the git / GitHub side.

**Reply by daaan** [reply]
Same here. First seen today:

> remote: Bypassed rule violations for refs/heads/master:
> remote:
> remote: - At least 1 approving review is required by reviewers with write access.

Commit went fine but raised doubts among developers.

**Reply by zaikunzhang** [reply]
> Same here. First seen today:
> 
> > remote: Bypassed rule violations for refs/heads/master:
> > remote:
> > remote: - At least 1 approving review is required by reviewers with write access.
> 
> Commit went fine but raised doubts among developers.

Yes, the commit still works despite the warning. In addition, no notification or whatever is raised on the GitHub repo, so that I cannot "review the commit" even if I want to. Everything is just normal except for this warning.  

Good to know that I am not alone.

**Reply by leoalenc** [reply]
> Thank you @airtower-luna . I see it in the same way as you.
> 
> However, it has never happended before the SSH key incedent of GitHub. In addition, [someone else](https://github.com/orgs/community/discussions/50878#discussioncomment-5428612) experiences something similar due to the change of the SSH key. That is why I thought they were related.
> 
> > I assume that you have admin rights to the repository and the rule is configured to allow admins to ignore it.
> 
> Right. **I am the only owner, administrator, and contributor of the repo.** The warning is not wrong, but it seems to have been suppressed by default before. The question is, how and why is it turned on now (starting two days ago)?
> 
> Thanks.
> 
> Remark: I found three similar cases on the Internet.
> 
> * GitHub Community Discourse: https://github.com/orgs/community/discussions/50878#discussioncomment-5428612
> * StackExchange: https://stackoverflow.com/questions/75839192/github-at-least-1-approving-review-is-required-by-reviewers-with-write-access
> * GitHub code: https://github.com/RikaShiro/FYP22070/blob/a15eae1e9a0cdbb1b2d7c14d640549ada6c8f17a/FYP/posts/20230320-20230326.txt#L15
> 
> All of them started to occur two days ago. I speculate that they are caused by some recent changes on the git / GitHub side.

I have experiencing the same problem since a few days ago, exactly the same way as reported by @zaikunzhang.

**Reply by Nozibul** [reply]
https://docs.github.com/en/code-security/secret-scanning/configuring-secret-scanning-for-your-repositories#excluding-directories-from-secret-scanning-alerts-for-users

follow the step and solve the issue

**Comment by zaikunzhang** [comment]
Is there any possibility that what we have observed is just the preclusion of the severe incident  https://github.com/orgs/community/discussions/15823 ?

**Reply by daaan** [reply]
We had the same thought over here. Hope it gets resolved soon.

**Comment by jsoref** [comment, marked as answer]
What is your goal in removing?

Do you want to be rejected if your push doesn't have the required statuses?

Do you want to be able to bypass and just not see the warning that you're bypassing?

**Reply by zaikunzhang** [reply]
Hi @jsoref . Thank you for looking at this. 

I hope things work as before. As far as I understand, everything was working as "bypass and just not see the warning that you're bypassing" until last Friday. (In the branch protection rule, I already specified that I can bypass the reviewing rule).

**Reply by jsoref** [reply]
Technically you could use `core.sshCommand`, `$GIT_SSH_COMMAND`, or `$GIT_SSH`. You would need to write a script/program that takes its arguments and runs ssh with them, passes input to the normal ssh program, filters output according to your desires, and otherwise returns output.

That said, if you're asking, it's likely you aren't ready to do this. I'd suggest taking the standard approach and just living with the additional output.

**Reply by zaikunzhang** [reply]
Even though what you suggest may be technical to implement, I understand it --- just wrap the original `git` under some cover, and filter out the messages that I do not want to see. 

Whether is this a good strategy or not, it is still interesting to ask what makes this message start to pop out a few days ago and why it had been suppressed by default before.

**Reply by jsoref** [reply]
Clearly what made it happen was "progress". Some GitHub manager realized that people weren't aware that they were bypassing rules and decided that the best (really only) place to inform users of which rules they were bypassing was here, so they added code to do so.

These kinds of improvements make auditors happy. I highly doubt they'd be removed/configurable.

Anyway. I think that should give you the information you need to decide whether it's worth spending engineering effort to discard the output or just accept that output.

Keep in mind that the channel they're using can always grow additional messaging.

**Reply by zaikunzhang** [reply]
Thank you @jsoref  for the comments. Sounds reasonable. 

Just in case, I did not mean to "remove" anything ...

**Reply by chkoreff** [reply]
It's not that I want to bypass the warning, I want to fix what it's warning about.  It's telling me that:

"At least 1 approving review is required by reviewers with write access."

OK, I don't know of anyone with write access beside me, but in any case, how do I obtain the required review?  What is a review anyway?

**Reply by jsoref** [reply]
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review

**Reply by airtower-luna** [reply]
@chkoreff, that message implies you have a branch protection rule that requires

* changes to go through PRs to get into the branch you pushed to, AND
* that someone else with write access (so not a random person) approves the PR before it gets merged (that's the review).

Obviously if you're the only one with write access to the project you can't fill that condition, but it might still make sense to keep the rule for outside contributors who's code you'll want to review before merging.

**Reply by jsoref** [reply]
@chkoreff: It looks like [`gh pr merge`](https://cli.github.com/manual/gh_pr_merge) `--admin` might be your friend.

**Comment by corwinn** [comment]
Same here:
`remote: Resolving deltas: 100% (3/3), completed with 3 local objects.`
`remote: Bypassed rule violations for refs/heads/main:`
`remote:`
`remote: - At least 1 approving review is required by reviewers with write access.`
`remote:`
This doesn't have objective reasoning, yet. It started happening out of the blue.
Either I failed understanding branch protection, or something else is at play.

**Reply by zaikunzhang** [reply]
It seems that GitHub is making progress in making all of us more vigilant about branch protection.

**Comment by zaikunzhang** [comment]
Problem solved. According to GitHub support: 

"it is expected you'll see those warnings now! We've just rolled that out as a feature ..."

So @jsoref  is correct. 

Thank everyone for the discussion.

**Comment by zaikunzhang** [comment]
See https://github.blog/changelog/2023-03-29-info-messages-for-bypassing-branch-protections-in-git/

**Comment by patrick-knight** [comment]
Thanks @zaikunzhang for posting the change log here.


This is a new behavior we are rolling out to let an admin know they are committing to a repo with branch protections in place and will bypass those rules based on their permission. 

Here's the [change log ](https://github.blog/changelog/2023-03-29-info-messages-for-bypassing-branch-protections-in-git/)we shipped about this update.

**Reply by niij** [reply]
@patrick-knight did this _used_ to block users from pushing even if they had "bypass" permission? I want branch protection to block accidental pushes from the command line, but to still allows users to merge a PR if they click the scary red "bypass" checkbox.

Is there a way to block pushes to main, but still allow merging PRs who are currently violating the branch protection rules in the Github UI?

**Reply by jsoref** [reply]
My understanding is that users silently bypassed and now they noisily bypass.

My guess is that you can't achieve what you want.

What you want could be achieved by changing how your workflow works. Right now you have a system where checks A, B and C are required.

Instead of doing that, you could have a new check R which is required and have it check the status of a bunch of things or the presence of a label "bypass-checks". And then when you need to bypass checks, you add the label (or have something add it...).

**Reply by patrick-knight** [reply]
There is a bypass via PR option that is available:

![GitHubScreenShot-Arc-24-01-2024-002473@2x](https://github.com/community/community/assets/7575792/8175a009-b452-485a-bd52-bf6bdaeaedc4)
 
 That will block command line pushes but allow PR merging.

**Reply by patrick-knight** [reply]
And no prior to this admins saw no warnings when they bypassed rules.

**Reply by jsoref** [reply]
Fwiw, you can use `gh pr merge --admin` to do the equivalent of the GUI override, but, indeed the bypass stuff does block pushes (pretty cool).

**Comment by bricksdont** [comment]
So it seems this is purely informational, but it breaks my mental workflow because every time I commit I'm under the impression that the commit push did not work. How to disable the message?

**Reply by jsoref** [reply]
https://github.com/orgs/community/discussions/51047#discussioncomment-5441351

**Comment by domdfcoding** [comment]
Before the `Require a pull request before merging` option was added` (new to me) approvals were required on PRs but PRs themselves weren't required. There's not now an option for this, as disabling `Require a pull request` disables the requirement for PRs to have reviews. 

Am I missing something in the muddle of branch protection options, as that would make the annoying message go away for me?

**Reply by patrick-knight** [reply]
You can set the reviewer count to `0`

**Comment by tomjoht** [comment]
The wording of the message makes it seem like an error. Instead of this:

```
remote: Bypassed rule violations for refs/heads/main:
remote: 
remote: - Commits must have valid signatures.
```

Change it to something like this:

```
Authorizing commit based on valid signature
```

Right now, the commit message reads like it says "Rule violation - commit must have valid signature". Do you really want to show that message to everyone, making them feel the error? I just removed my GPG keys and readded them thinking the originals had expired, but I was confused b/c the commits still went through. Pls consider the usability of the messaging more.

**Reply by patrick-knight** [reply]
The intent of the message is that there is a rule in place on a repository, that is typically applied but you have an exemption. 

In the situation you shared that message is present because a valid signature was not used and you had permission to push bypassing the signature rule.

**Reply by jsoref** [reply]
That's correct, but @tomjoht is right, the message does a bad job of explaining what's going on.

Consider:
```
remote: Encountered rule violations for refs/heads/main:
remote: 
remote: - Commits must have valid signatures.
remote: 
remote: These rules were ignored because the pusher is an admin.
```

**Comment by Challanger524** [comment]
@patrick-knight, this microsoft-usual-style _BS_ "feature" is **confusing as hell**.
If I manually setup bypasses to myself - I DO EXPECT to not to be bothered by weirdly-formed messages the way it seems that push is declined despite nothing being declined 🤯.
Microsoft: lets make worse by adding a "feature" nobody asked for - Absolutely!
